File: hdf-diff/src/normalize.ts
/**
 * Normalize InSpec exec-json v1 format to HDF v2 structure for diffing.
 *
 * V1 uses: profiles[].controls[].results[].code_desc, source_location, start_time
 * V2 uses: baselines[].requirements[].results[].codeDesc, sourceLocation, startTime
 *
 * This module detects v1 documents and converts them in-memory to v2 shape
 * so the diff engine only needs to handle one format.
 */

interface V1Result {
  status: string;
  code_desc?: string;
  codeDesc?: string;
  run_time?: number;
  runTime?: number;
  start_time?: string;
  startTime?: string;
  message?: string;
  [key: string]: unknown;
}

interface V1Control {
  id: string;
  title?: string;
  desc?: string;
  impact: number;
  tags?: Record<string, unknown>;
  refs?: unknown[];
  code?: string;
  source_location?: { ref?: string; line?: number };
  sourceLocation?: { ref?: string; line?: number };
  results?: V1Result[];
  [key: string]: unknown;
}

interface V1Profile {
  name: string;
  title?: string;
  version?: string;
  sha256?: string;
  controls?: V1Control[];
  groups?: unknown[];
  supports?: unknown[];
  attributes?: unknown[];
  [key: string]: unknown;
}

/**
 * Detect whether a document is v1 (InSpec exec-json) format.
 * V1 has `profiles` at the top level; v2 has `baselines`.
 */
export function isV1Format(doc: Record<string, unknown>): boolean {
  return Array.isArray(doc['profiles']) && !Array.isArray(doc['baselines']);
}

/**
 * Normalize a document to v2-like structure. If already v2, returns as-is.
 * If v1, converts profiles→baselines, controls→requirements, and snake_case→camelCase.
 */
export function normalizeToV2(doc: Record<string, unknown>): Record<string, unknown> {
  if (!isV1Format(doc)) {
    return doc;
  }

  const profiles = doc['profiles'] as V1Profile[];
  const baselines = profiles.map(normalizeProfile);

  // Preserve timestamp from statistics if available
  const statistics = doc['statistics'] as Record<string, unknown> | undefined;

  return {
    baselines,
    statistics,
    timestamp: doc['timestamp'],
  };
}

function normalizeProfile(profile: V1Profile): Record<string, unknown> {
  const controls = profile.controls ?? [];
  const requirements = controls.map(normalizeControl);

  return {
    name: profile.name,
    title: profile.title,
    version: profile.version,
    checksum: profile.sha256 ? { algorithm: 'sha256', value: profile.sha256 } : undefined,
    groups: profile.groups ?? [],
    supports: profile.supports ?? [],
    inputs: profile.attributes ?? [],
    requirements,
  };
}

function normalizeControl(control: V1Control): Record<string, unknown> {
  const v1Results = control.results ?? [];
  const results = v1Results.map(normalizeResult);

  return {
    id: control.id,
    title: control.title,
    descriptions: control.desc
      ? [{ label: 'default', data: control.desc }]
      : [],
    impact: control.impact,
    tags: control.tags ?? {},
    refs: control.refs ?? [],
    code: control.code,
    sourceLocation: control.source_location ?? control.sourceLocation,
    results,
  };
}

/**
 * Map InSpec v1 result status values to HDF v2 Result_Status enum values.
 *
 * InSpec v1 uses: "passed", "failed", "skipped", "error"
 * HDF v2 uses: "passed", "failed", "notApplicable", "notReviewed", "error"
 *
 * "skipped" in InSpec means the test was not executed (typically because a
 * `describe.one_of` condition was not met, or `only_if` excluded it).
 * This maps to "notReviewed" in HDF v2 — the requirement was not assessed.
 */
function normalizeResultStatus(v1Status: string): string {
  switch (v1Status) {
    case 'skipped':
      return 'notReviewed';
    default:
      return v1Status;
  }
}

/**
 * Normalize a timestamp string to RFC 3339 / ISO 8601 date-time format.
 *
 * InSpec v1 uses formats like "2017-09-22 14:12:15 -0400" which are not valid
 * RFC 3339. This function attempts to parse and re-format such timestamps.
 * If parsing fails, the original string is returned as-is.
 */
function normalizeTimestamp(timestamp: string): string {
  // Already valid ISO 8601 (contains 'T')
  if (timestamp.includes('T')) {
    return timestamp;
  }

  // Try to parse InSpec format: "YYYY-MM-DD HH:MM:SS +HHMM"
  const parsed = new Date(timestamp);
  if (!isNaN(parsed.getTime())) {
    return parsed.toISOString();
  }

  return timestamp;
}

function normalizeResult(result: V1Result): Record<string, unknown> {
  const rawStartTime = result.start_time ?? result.startTime ?? '';
  const normalized: Record<string, unknown> = {
    status: normalizeResultStatus(result.status),
    codeDesc: result.code_desc ?? result.codeDesc ?? '',
    startTime: rawStartTime ? normalizeTimestamp(rawStartTime) : rawStartTime,
  };

  // Only include optional fields when they have values
  const runTime = result.run_time ?? result.runTime;
  if (runTime !== undefined) {
    normalized['runTime'] = runTime;
  }
  if (result.message !== undefined) {
    normalized['message'] = result.message;
  }

  return normalized;
}

File: hdf-diff/go/normalize.go
// Package diff provides legacy InSpec exec-json to HDF normalization.
//
// Legacy format uses profiles[].controls[] with snake_case fields.
// Current HDF format uses baselines[].requirements[] with camelCase fields.
package diff

import (
	"encoding/json"
	"fmt"

	hdfutil "github.com/mitre/hdf-libs/hdf-utilities/go"
	hdf "github.com/mitre/hdf-schema"
)

// IsV1Format detects whether a JSON document is legacy InSpec exec-json format.
// Legacy format has "profiles" at top level; current HDF has "baselines".
func IsV1Format(data map[string]any) bool {
	_, hasProfiles := data["profiles"]
	_, hasBaselines := data["baselines"]

	if !hasProfiles {
		return false
	}

	// Check that profiles is actually an array
	if _, ok := data["profiles"].([]any); !ok {
		return false
	}

	return !hasBaselines
}

// ToV2 converts a legacy InSpec exec-json document to an HdfResults struct.
// If the input is already current HDF format, parses directly. If legacy,
// converts profiles→baselines, controls→requirements, snake_case→camelCase.
// The returned warnings slice contains messages for skipped profiles/controls
// during legacy conversion. For current-format passthrough, warnings is nil.
func ToV2(data []byte) (hdf.HDFResults, []string, error) {
	// First, parse into a generic map to detect format
	var raw map[string]any
	if err := json.Unmarshal(data, &raw); err != nil {
		return hdf.HDFResults{}, nil, err
	}

	if !IsV1Format(raw) {
		// Current HDF format: parse directly into typed struct
		result, err := hdf.UnmarshalHDFResults(data)
		return result, nil, err
	}

	// Legacy format: convert to current HDF structure
	return convertV1ToV2(raw)
}

func convertV1ToV2(raw map[string]any) (hdf.HDFResults, []string, error) {
	result := hdf.HDFResults{}
	var warnings []string

	// Convert profiles -> baselines
	profiles, _ := raw["profiles"].([]any)
	baselines := make([]hdf.EvaluatedBaseline, 0, len(profiles))
	for i, p := range profiles {
		profileMap, ok := p.(map[string]any)
		if !ok {
			warnings = append(warnings, fmt.Sprintf("skipped profile at index %d: not a JSON object", i))
			continue
		}
		baseline, profileWarnings := normalizeProfile(profileMap)
		warnings = append(warnings, profileWarnings...)
		baselines = append(baselines, baseline)
	}
	result.Baselines = baselines

	// Convert statistics
	if stats, ok := raw["statistics"].(map[string]any); ok {
		s := normalizeStatistics(stats)
		result.Statistics = &s
	}

	// Preserve timestamp
	if ts, ok := raw["timestamp"].(string); ok && ts != "" {
		parsed := hdfutil.ParseTimestamp(ts)
		if !parsed.IsZero() {
			result.Timestamp = &parsed
		}
	}

	return result, warnings, nil
}

func normalizeProfile(profile map[string]any) (hdf.EvaluatedBaseline, []string) {
	baseline := hdf.EvaluatedBaseline{}
	var warnings []string

	baseline.Name = getString(profile, "name")

	if title, ok := profile["title"].(string); ok {
		baseline.Title = &title
	}
	if version, ok := profile["version"].(string); ok {
		baseline.Version = &version
	}

	// sha256 -> integrity
	if sha, ok := profile["sha256"].(string); ok && sha != "" {
		alg := hdf.Sha256
		baseline.Integrity = &hdf.Integrity{
			Algorithm: &alg,
			Checksum:  &sha,
		}
	}

	// Groups
	baseline.Groups = normalizeGroups(profile)

	// Supports
	baseline.Supports = normalizeSupports(profile)

	// Inputs: legacy "attributes" are unstructured maps that don't conform to
	// the typed Input schema. Store in extensions to preserve the data.
	if attrs := normalizeInputs(profile); len(attrs) > 0 {
		if baseline.Extensions == nil {
			baseline.Extensions = make(map[string]interface{})
		}
		baseline.Extensions["legacyAttributes"] = attrs
	}

	// Controls -> Requirements
	controls, _ := profile["controls"].([]any)
	reqs := make([]hdf.EvaluatedRequirement, 0, len(controls))
	for i, c := range controls {
		controlMap, ok := c.(map[string]any)
		if !ok {
			warnings = append(warnings, fmt.Sprintf("profile %q: skipped control at index %d: not a JSON object", baseline.Name, i))
			continue
		}
		reqs = append(reqs, normalizeControl(controlMap))
	}
	baseline.Requirements = reqs

	return baseline, warnings
}

func normalizeControl(control map[string]any) hdf.EvaluatedRequirement {
	req := hdf.EvaluatedRequirement{}

	req.ID = getString(control, "id")

	if title, ok := control["title"].(string); ok {
		req.Title = &title
	}

	// desc -> descriptions array
	if desc, ok := control["desc"].(string); ok && desc != "" {
		req.Descriptions = []hdf.Description{
			{Label: "default", Data: desc},
		}
	} else {
		req.Descriptions = []hdf.Description{}
	}

	req.Impact = getFloat(control, "impact")

	// Tags
	if tags, ok := control["tags"].(map[string]any); ok {
		req.Tags = tags
	} else {
		req.Tags = map[string]any{}
	}

	// Refs
	req.Refs = normalizeRefs(control)

	// Code
	if code, ok := control["code"].(string); ok {
		req.Code = &code
	}

	// source_location or sourceLocation
	sl := normalizeSourceLocation(control)
	req.SourceLocation = &sl

	// Results
	results, _ := control["results"].([]any)
	reqResults := make([]hdf.RequirementResult, 0, len(results))
	for _, r := range results {
		resultMap, ok := r.(map[string]any)
		if !ok {
			continue
		}
		reqResults = append(reqResults, normalizeResult(resultMap))
	}
	req.Results = reqResults

	return req
}

func normalizeResult(result map[string]any) hdf.RequirementResult {
	r := hdf.RequirementResult{}

	// Status mapping: "skipped" -> "notReviewed"
	status := normalizeResultStatus(getString(result, "status"))
	r.Status = hdf.ResultStatus(status)

	// code_desc or codeDesc
	r.CodeDesc = getStringFallback(result, "code_desc", "codeDesc")

	// start_time or startTime -> parsed time
	rawStartTime := getStringFallback(result, "start_time", "startTime")
	if rawStartTime != "" {
		r.StartTime = hdfutil.ParseTimestamp(rawStartTime)
	}
	// If empty or missing, StartTime stays as zero value

	// run_time or runTime (optional)
	if rt, ok := getFloatOptional(result, "run_time"); ok {
		r.RunTime = &rt
	} else if rt, ok := getFloatOptional(result, "runTime"); ok {
		r.RunTime = &rt
	}

	// message (optional)
	if msg, ok := result["message"].(string); ok {
		r.Message = &msg
	}

	return r
}

// normalizeResultStatus maps legacy InSpec status values to HDF ResultStatus values.
func normalizeResultStatus(status string) string {
	switch status {
	case "skipped":
		return "notReviewed"
	default:
		return status
	}
}

func normalizeSourceLocation(control map[string]any) hdf.SourceLocation {
	sl := hdf.SourceLocation{}

	// Try snake_case first, then camelCase
	var loc map[string]any
	if v, ok := control["source_location"].(map[string]any); ok {
		loc = v
	} else if v, ok := control["sourceLocation"].(map[string]any); ok {
		loc = v
	}

	if loc != nil {
		if ref, ok := loc["ref"].(string); ok {
			sl.Ref = &ref
		}
		if line, ok := loc["line"].(float64); ok {
			sl.Line = &line
		}
	}

	return sl
}

func normalizeStatistics(stats map[string]any) hdf.Statistics {
	s := hdf.Statistics{}
	if d, ok := stats["duration"].(float64); ok {
		s.Duration = &d
	}
	return s
}

func normalizeGroups(profile map[string]any) []hdf.RequirementGroup {
	groups, ok := profile["groups"].([]any)
	if !ok {
		return []hdf.RequirementGroup{}
	}
	result := make([]hdf.RequirementGroup, 0, len(groups))
	for _, g := range groups {
		if gMap, ok := g.(map[string]any); ok {
			rg := hdf.RequirementGroup{
				ID: getString(gMap, "id"),
			}
			if title, ok := gMap["title"].(string); ok {
				rg.Title = &title
			}
			result = append(result, rg)
		}
	}
	return result
}

func normalizeSupports(profile map[string]any) []hdf.SupportedPlatform {
	supports, ok := profile["supports"].([]any)
	if !ok {
		return []hdf.SupportedPlatform{}
	}
	result := make([]hdf.SupportedPlatform, 0, len(supports))
	for _, s := range supports {
		if sMap, ok := s.(map[string]any); ok {
			sp := hdf.SupportedPlatform{}
			if p, ok := sMap["platform"].(string); ok {
				sp.Platform = &p
			}
			if pn, ok := sMap["platformName"].(string); ok {
				sp.PlatformName = &pn
			}
			result = append(result, sp)
		}
	}
	return result
}

func normalizeInputs(profile map[string]any) []map[string]any {
	attrs, ok := profile["attributes"].([]any)
	if !ok {
		return []map[string]any{}
	}
	result := make([]map[string]any, 0, len(attrs))
	for _, a := range attrs {
		if aMap, ok := a.(map[string]any); ok {
			result = append(result, aMap)
		}
	}
	return result
}

func normalizeRefs(control map[string]any) []hdf.Reference {
	refs, ok := control["refs"].([]any)
	if !ok {
		return []hdf.Reference{}
	}
	result := make([]hdf.Reference, 0, len(refs))
	for _, r := range refs {
		ref := hdf.Reference{}
		switch v := r.(type) {
		case string:
			// Plain string ref: wrap in Ref.String
			ref.Ref = &hdf.Ref{String: &v}
		case map[string]any:
			// Object ref: extract "ref", "url", "uri" fields
			if refStr, ok := v["ref"].(string); ok {
				ref.Ref = &hdf.Ref{String: &refStr}
			}
			if urlStr, ok := v["url"].(string); ok {
				ref.URL = &urlStr
			}
			if uriStr, ok := v["uri"].(string); ok {
				ref.URI = &uriStr
			}
		default:
			// Unknown format: add empty Reference to preserve count
		}
		result = append(result, ref)
	}
	return result
}

// Helper functions for safe type extraction from map[string]any.

func getString(m map[string]any, key string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return ""
}

func getStringFallback(m map[string]any, snakeKey, camelKey string) string {
	if v, ok := m[snakeKey].(string); ok {
		return v
	}
	if v, ok := m[camelKey].(string); ok {
		return v
	}
	return ""
}

func getFloat(m map[string]any, key string) float64 {
	if v, ok := m[key].(float64); ok {
		return v
	}
	return 0
}

func getFloatOptional(m map[string]any, key string) (float64, bool) {
	v, ok := m[key].(float64)
	return v, ok
}
